universalModule/software/RoFIDriver/src/rofi_hal.hpp

```diff
@@ -0,0 +1,344 @@
+    Internal = 0,
+    External = 1
+};
+
+/**
+ * \brief Connector state descriptor
+ */
+struct ConnectorState {
+    ConnectorPosition position = ConnectorPosition::Retracted;
+    bool internal = false; ///< Is internal power bus connected to the connector?
+    bool external = false; ///< Is the external power bus connected to the connector?
+    bool connected = false; ///< Is there a mating side connected?
+    ConnectorOrientation orientation = ConnectorOrientation::North;
+    float internalVoltage, internalCurrent;
+    float externalVoltage, externalCurrent;
+};
+
+enum ConnectorEvent {
+    Connected = 0,
+    Disconnected = 1
+};
+
+/**
+ * \brief Proxy for controlling a single connector of RoFI
+ *
+ * Joint cannot be instantiated on its own; you can obtain its instance from
+ * RoFI::getConnector().
+ *
+ * Since the class is only a proxy it can be safely copied and passed around.
+ */
+class Connector {
+public:
+    /**
+     *  \brief Interface of the actual implementation - either physical local,
+     *  physical remote or a simulation one.
+     *
+     *  To provide an implementation, simply inherit from this class.
+     *
+     *  For method description, see description of Connector.
+     */
+    class Implementation {
+    public:
+        virtual ConnectorState getState() const = 0;
+        virtual void connect() = 0;
+        virtual void disconnect() = 0;
+        virtual void onConnectorEvent(
+            std::function< void ( Connector, ConnectorEvent ) > callback ) = 0;
+        virtual void onPacket( std::function< void( Connector,
+            uint16_t contentType, PBuf ) > callback ) = 0;
+        virtual void send( uint16_t contentType, PBuf packet ) = 0;
+        virtual void connectPower( ConnectorLine ) = 0;
+        virtual void disconnectPower( ConnectorLine ) = 0;
+    };
+
+    /**
+     * \brief Get connector state
+     */
+    ConnectorState getState() const { return _impl->getState(); }
+
+    /**
+     * \brief Expand the connector to be ready to accept connection.
+     *
+     * This action does not establishes connection, register a callback via
+     * Connector::onConnectorEvent().
+     */
+    void connect() { _impl->connect(); }
+
+    /**
+     * \brief Retract the connector.
+     *
+     * Does not releases connection immediately - after the connection is broken
+     * physically, calls callback registered via Connector::onConnectorEvent().
+     */
+    void disconnect() { _impl->disconnect(); }
+
+    /**
+     * \brief Register callback for connector events
+     */
+    void onConnectorEvent( std::function< void ( Connector, ConnectorEvent ) > callback ) {
+        _impl->onConnectorEvent( callback );
+    }
+
+    /**
+     * \brief Register callback for incoming packets.
+     *
+     * The callback takes connector which received the packet, contentType and
+     * the packet.
+     */
+    void onPacket( std::function< void( Connector, uint16_t, PBuf ) > callback ) {
```

Comment:
As we discussed - call all callbacks given (and add this information to the docs)